In [ ]:
from azureml.core import Workspace, Experiment

ws = Workspace.from_config()
#ws = Workspace.get(name="udacity-project")
exp = Experiment(workspace=ws, name="udacity-project")

print('Workspace name: ' + ws.name, 
      'Azure region: ' + ws.location, 
      'Subscription id: ' + ws.subscription_id, 
      'Resource group: ' + ws.resource_group, sep = '\n')

In [2]:
from azureml.core.compute import ComputeTarget, AmlCompute
from azureml.core.compute_target import ComputeTargetException

# Choose a name for your CPU cluster
cpu_cluster_name = "cluster"

# Verify that cluster does not exist already
try:
    compute_target = ComputeTarget(workspace=ws, name=cpu_cluster_name)
    print('Found existing cluster, use it.')
except ComputeTargetException:
    compute_config = AmlCompute.provisioning_configuration(vm_size='STANDARD_D3_v2',
                                                           max_nodes=10)
    compute_target = ComputeTarget.create(ws, cpu_cluster_name, compute_config)

compute_target.wait_for_completion(show_output=True)
    
# For a more detailed view of current AmlCompute status, use get_status().

Found existing cluster, use it.
Succeeded
AmlCompute wait for completion finished

Minimum number of nodes requested have been provisioned


In [3]:
from azureml.widgets import RunDetails
from azureml.train.sklearn import SKLearn
from azureml.train.hyperdrive.run import PrimaryMetricGoal
from azureml.train.hyperdrive.policy import BanditPolicy
from azureml.train.hyperdrive.sampling import RandomParameterSampling
from azureml.train.hyperdrive.runconfig import HyperDriveConfig
from azureml.train.hyperdrive.parameter_expressions import uniform
from azureml.train.hyperdrive.parameter_expressions import choice
import os

# Specify parameter sampler
ps = RandomParameterSampling( {
    "--C": uniform(0.0, 1.0),
    "--max-iter": choice(100, 200, 400, 800)
    }
)

# Specify a Policy
policy = BanditPolicy(evaluation_interval=2, slack_factor=0.1)

if "training" not in os.listdir():
    os.mkdir("./training")

# Create a SKLearn estimator for use with train.py
pip_packages = [
                'azureml-core==1.17.0', 'azureml-telemetry==1.17.0', 'azureml-dataprep==2.4.2',
                'joblib==0.14.1', 'pandas==1.0.0', 'sklearn-pandas==2.0.2' 
               ]

est = SKLearn(source_directory='.',
              #script_params=script_params,
              compute_target=compute_target,
              entry_script='train.py',
              pip_packages=pip_packages,
              conda_packages=['scikit-learn==0.22.2.post1'])

# Create a HyperDriveConfig using the estimator, hyperparameter sampler, and policy.
hyperdrive_config = HyperDriveConfig(
    estimator=est,
    hyperparameter_sampling=ps,
    policy=policy,
    primary_metric_name='Accuracy',
    primary_metric_goal=PrimaryMetricGoal.MAXIMIZE,
    max_total_runs=20,
    max_concurrent_runs=4)

WARNING - You have specified to install packages in your run. Note that you have overridden Azure ML's installation of the following packages: ['joblib', 'scikit-learn']. We cannot guarantee image build will succeed.


In [4]:
# Submit your hyperdrive run to the experiment and show run details with the widget.
hyperdrive_run = exp.submit(hyperdrive_config)

WARNING - If 'script' has been provided here and a script file name has been specified in 'run_config', 'script' provided in ScriptRunConfig initialization will take precedence.


In [5]:
from azureml.widgets import RunDetails
RunDetails(hyperdrive_run).show()

_HyperDriveWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO'…

{'runId': 'HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b',
 'target': 'cluster',
 'status': 'Completed',
 'startTimeUtc': '2020-11-09T17:58:40.620801Z',
 'endTimeUtc': '2020-11-09T18:17:42.436353Z',
 'properties': {'primary_metric_config': '{"name": "Accuracy", "goal": "maximize"}',
  'resume_from': 'null',
  'runTemplate': 'HyperDrive',
  'azureml.runsource': 'hyperdrive',
  'platform': 'AML',
  'ContentSnapshotId': '69be2a33-5216-4e0c-9e10-98d92c233130',
  'score': '0.9128983308042489',
  'best_child_run_id': 'HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b_6',
  'best_metric_status': 'Succeeded'},
 'inputDatasets': [],
 'outputDatasets': [],
 'logFiles': {'azureml-logs/hyperdrive.txt': 'https://demoentws5367325393.blob.core.windows.net/azureml/ExperimentRun/dcid.HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b/azureml-logs/hyperdrive.txt?sv=2019-02-02&sr=b&sig=iqOa0GvdbNvpFGxNh2gvSXxv5r2bk%2BQKpjO4MXi5%2FkM%3D&st=2020-11-09T18%3A07%3A50Z&se=2020-11-10T02%3A17%3A50Z&sp=r'}}

In [8]:
hyperdrive_run.wait_for_completion(show_output=False)

{'runId': 'HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b',
 'target': 'cluster',
 'status': 'Completed',
 'startTimeUtc': '2020-11-09T17:58:40.620801Z',
 'endTimeUtc': '2020-11-09T18:17:42.436353Z',
 'properties': {'primary_metric_config': '{"name": "Accuracy", "goal": "maximize"}',
  'resume_from': 'null',
  'runTemplate': 'HyperDrive',
  'azureml.runsource': 'hyperdrive',
  'platform': 'AML',
  'ContentSnapshotId': '69be2a33-5216-4e0c-9e10-98d92c233130',
  'score': '0.9128983308042489',
  'best_child_run_id': 'HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b_6',
  'best_metric_status': 'Succeeded'},
 'inputDatasets': [],
 'outputDatasets': [],
 'logFiles': {'azureml-logs/hyperdrive.txt': 'https://demoentws5367325393.blob.core.windows.net/azureml/ExperimentRun/dcid.HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b/azureml-logs/hyperdrive.txt?sv=2019-02-02&sr=b&sig=iqOa0GvdbNvpFGxNh2gvSXxv5r2bk%2BQKpjO4MXi5%2FkM%3D&st=2020-11-09T18%3A07%3A50Z&se=2020-11-10T02%3A17%3A50Z&sp=r'}}

In [9]:
# Get your best run and save the model from that run.
best_run = hyperdrive_run.get_best_run_by_primary_metric()

print(best_run.get_details()['runId'])
print(best_run.get_details()['runDefinition']['arguments'])
print(best_run.get_metrics())

HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b_6
['--C', '0.45085242957194405', '--max-iter', '100']
{'Regularization Strength:': 0.45085242957194405, 'Max iterations:': 100, 'Accuracy': 0.9128983308042489}


### Model parameters
- RunID: HD_d61d0d18-d18b-4535-b24a-1336cfa97f8b_6
- Regularization Strength (C): 0.45085242957194405
- Max iterations: 100
- Accuracy: 0.9128983308042489

In [10]:
import joblib

os.makedirs('./training', exist_ok=True)

best_run.download_file(name='outputs/bankmarketing_hyperdrive.pkl',
                       output_file_path='training/bankmarketing_hyperdrive.pkl')

# AutoML Experiment

In [ ]:
# from azureml.core import Workspace, Experiment

# ws = Workspace.from_config()
# #ws = Workspace.get(name="udacity-project")
# exp = Experiment(workspace=ws, name="udacity-project")

# print('Workspace name: ' + ws.name, 
#       'Azure region: ' + ws.location, 
#       'Subscription id: ' + ws.subscription_id, 
#       'Resource group: ' + ws.resource_group,
#       'Azure ML SDK: ' + azureml.core.VERSION, sep = '\n')

In [11]:
# Create TabularDataset using TabularDatasetFactory
# Data is available at: 
# "https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv"

from azureml.core import Dataset

# create a FileDataset from image and label files behind public web urls
web_path = 'https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv'

ds = Dataset.Tabular.from_delimited_files(path=web_path)

In [12]:
ds.take(5).to_pandas_dataframe()

age          job  marital    education  default housing loan    contact  \
0   57   technician  married  high.school       no      no  yes   cellular   
1   55      unknown  married      unknown  unknown     yes   no  telephone   
2   33  blue-collar  married     basic.9y       no      no   no   cellular   
3   36       admin.  married  high.school       no      no   no  telephone   
4   27    housemaid  married  high.school       no     yes   no   cellular   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         1      failure         -1.8   
1   may         thu  ...         2    999         0  nonexistent          1.1   
2   may         fri  ...         1    999         1      failure         -1.8   
3   jun         fri  ...         4    999         0  nonexistent          1.4   
4   jul         fri  ...         2    999         0  nonexistent          1.4   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          92.893          -46.2      1.299       5099.1  no  
1          93.994          -36.4      4.860       5191.0  no  
2          92.893          -46.2      1.313       5099.1  no  
3          94.465          -41.8      4.967       5228.1  no  
4          93.918          -42.7      4.963       5228.1  no  

[5 rows x 21 columns]

In [13]:
from train import clean_data

x, y = clean_data(ds)

In [19]:
import pandas as pd

# Append the target column to features
train_df = pd.concat([x, y], axis=1, sort=False)
train_df

age  marital  default  housing  loan  month  day_of_week  duration  \
0       57        1        0        0     1      5            1       371   
1       55        1        0        1     0      5            4       285   
2       33        1        0        0     0      5            5        52   
3       36        1        0        0     0      6            5       355   
4       27        1        0        1     0      7            5       189   
...    ...      ...      ...      ...   ...    ...          ...       ...   
32945   56        1        0        0     1      7            1       116   
32946   37        1        0        0     1      7            5        69   
32947   26        0        0        0     0      5            2       135   
32948   31        0        0        0     0      4            1       386   
32949   39        1        0        0     0      8            4       179   

       campaign  pdays  ...  contact_telephone  education_basic.4y  \
0             1    999  ...                  0                   0   
1             2    999  ...                  1                   0   
2             1    999  ...                  0                   0   
3             4    999  ...                  1                   0   
4             2    999  ...                  0                   0   
...         ...    ...  ...                ...                 ...   
32945         1    999  ...                  0                   1   
32946         7    999  ...                  0                   0   
32947         4    999  ...                  0                   0   
32948         1    999  ...                  0                   0   
32949         1    999  ...                  0                   1   

       education_basic.6y  education_basic.9y  education_high.school  \
0                       0                   0                      1   
1                       0                   0                      0   
2                       0                   1                      0   
3                       0                   0                      1   
4                       0                   0                      1   
...                   ...                 ...                    ...   
32945                   0                   0                      0   
32946                   0                   0                      0   
32947                   0                   0                      0   
32948                   0                   1                      0   
32949                   0                   0                      0   

       education_illiterate  education_professional.course  \
0                         0                              0   
1                         0                              0   
2                         0                              0   
3                         0                              0   
4                         0                              0   
...                     ...                            ...   
32945                     0                              0   
32946                     0                              0   
32947                     0                              0   
32948                     0                              0   
32949                     0                              0   

       education_university.degree  education_unknown  y  
0                                0                  0  0  
1                                0                  1  0  
2                                0                  0  0  
3                                0                  0  0  
4                                0                  0  0  
...                            ...                ... ..  
32945                            0                  0  0  
32946                            1                  0  0  
32947                            1                  0  0  
32948                            0       

In [15]:
from azureml.train.automl import AutoMLConfig

# Get a list of valid metrics for your given task
azureml.train.automl.utilities.get_primary_metrics('classification')

['accuracy',
 'average_precision_score_weighted',
 'norm_macro_recall',
 'precision_score_weighted',
 'AUC_weighted']

In [21]:
from azureml.train.automl import AutoMLConfig

# Set parameters for AutoMLConfig
automl_config = AutoMLConfig(
    experiment_timeout_minutes=30,
    task='classification',
    primary_metric='accuracy',
    training_data=train_df,
    label_column_name='y',
    n_cross_validations=10)

In [ ]:
# Submit your automl run
automl_run = exp.submit(automl_config, show_output = False)
automl_run

In [23]:
from azureml.widgets import RunDetails

RunDetails(automl_run).show()

_AutoMLWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO', 's…

In [24]:
automl_run.wait_for_completion(show_output=False)

{'runId': 'AutoML_07d0ba27-84b3-43fd-afab-d3c137b3e691',
 'target': 'local',
 'status': 'Completed',
 'startTimeUtc': '2020-11-09T18:41:35.594812Z',
 'endTimeUtc': '2020-11-09T19:15:43.888626Z',
 'properties': {'num_iterations': '1000',
  'training_type': 'TrainFull',
  'acquisition_function': 'EI',
  'primary_metric': 'accuracy',
  'train_split': '0',
  'acquisition_parameter': '0',
  'num_cross_validation': '10',
  'target': 'local',
  'AMLSettingsJsonString': '{"path":null,"name":"udacity-project","subscription_id":"bcbf34a7-1936-4783-8840-8f324c37f354","resource_group":"demo","workspace_name":"demo-ent-ws","region":"westeurope","compute_target":"local","spark_service":null,"azure_service":"Microsoft.AzureNotebookVM","_local_managed_run_id":null,"many_models":false,"pipeline_fetch_max_batch_size":1,"iterations":1000,"primary_metric":"accuracy","task_type":"classification","data_script":null,"validation_size":0.0,"n_cross_validations":10,"y_min":null,"y_max":null,"num_classes":2,"fea

In [25]:
# Retrieve and save your best automl model.

best_run, best_model = automl_run.get_output()

In [26]:
run_details = best_run.get_details()

model_details = {
    'RunID': [run_details['runId']],
    'Iteration': [run_details['properties']['iteration']],
    'Primary metric': [run_details['properties']['primary_metric']],
    'Score': [run_details['properties']['score']]
}

df = pd.DataFrame(model_details,
                  columns = ['RunID','Iteration','Primary metric','Score'],
                  index=[run_details['properties']['model_name']])
df

RunID Iteration  \
AutoML07d0ba27842  AutoML_07d0ba27-84b3-43fd-afab-d3c137b3e691_42        42   

                  Primary metric               Score  
AutoML07d0ba27842       accuracy  0.9180880121396054

In [40]:
# Print all the columns of a dataftame and don't truncate printed strings
pd.options.display.max_columns = None
pd.set_option('display.max_colwidth', -1)

In [53]:
# Show models and their hyper-parameters composing the Voting Ensamble
vc = best_model.steps[1][1]
print(vc.named_estimators)

{'0': Pipeline(memory=None,
         steps=[('maxabsscaler', MaxAbsScaler(copy=True)),
                ('lightgbmclassifier',
                 LightGBMClassifier(boosting_type='gbdt', class_weight=None,
                                    colsample_bytree=1.0,
                                    importance_type='split', learning_rate=0.1,
                                    max_depth=-1, min_child_samples=20,
                                    min_child_weight=0.001, min_split_gain=0.0,
                                    n_estimators=100, n_jobs=1, num_leaves=31,
                                    objective=None, random_state=None,
                                    reg_alpha=0.0, reg_lambda=0.0, silent=True,
                                    subsample=1.0, subsample_for_bin=200000,
                                    subsample_freq=0, verbose=-10))],
         verbose=False), '1': Pipeline(memory=None,
         steps=[('maxabsscaler', MaxAbsScaler(copy=True)),
                ('x

In [62]:
# Weights that the Voting Ensamble uses for each model respectively
print(vc.weights)

[0.09090909090909091, 0.09090909090909091, 0.18181818181818182, 0.18181818181818182, 0.09090909090909091, 0.09090909090909091, 0.18181818181818182, 0.09090909090909091]


### Model parameters
- RunID: AutoML_07d0ba27-84b3-43fd-afab-d3c137b3e691_42
- Accuracy: 0.9180880121396054

In [ ]:
best_run.download_file(name='outputs/model.pkl',
                       output_file_path='training/bankmarketing_automl.pkl')